src/callbacks/wandb_callbacks.py

```diff
@@ -41,18 +51,38 @@ def on_train_start(self, trainer, pl_module):
 
 
 class UploadCodeAsArtifact(Callback):
-    """Upload all *.py files to wandb as an artifact, at the beginning of the run."""
+    """Upload all code files to wandb as an artifact, at the beginning of the run."""
+
+    def __init__(self, code_dir: str, use_git: bool = True):
+        """
 
-    def __init__(self, code_dir: str):
+        :param code_dir:
+        :param use_git: if using git, then upload all files that are not ignored by git. if not using git, then upload all '*.py' file
+        """
         self.code_dir = code_dir
+        self.use_git = use_git
 
+    @rank_zero_only
     def on_train_start(self, trainer, pl_module):
         logger = get_wandb_logger(trainer=trainer)
         experiment = logger.experiment
 
         code = wandb.Artifact("project-source", type="code")
-        for path in glob.glob(os.path.join(self.code_dir, "**/*.py"), recursive=True):
-            code.add_file(path)
+
+        if self.use_git:
+            # get .git folder
+            # https://alexwlchan.net/2020/11/a-python-function-to-ignore-a-path-with-git-info-exclude/
+            git_dir_path = Path(subprocess.check_output(["git", "rev-parse", "--git-dir"]).strip().decode("utf8")).resolve()
+
+            for path in Path(self.code_dir).rglob('*'):
+                if (path.is_file()
+                        and (not is_relative_to(path, git_dir_path))  # ignore files in .git
```

Comment:
I replaced this line with:
```python
and not git_dir_path in path.parents
```
and it seems to still ignore `.git/` correctly (this way we don't need the `is_relative_to` method?)
Is this correct or am I missing something?

Reply:
Actually this should be faster:
```python
and not str(git_dir_path) in str(path)
```

Reply:
I think `str(git_dir_path) in str(path)` is potentially problematic in some edge cases. For example:
```python
# example .git directory
git_dir_path = Path('/home/foo/repo/.git')
# a path that is actually not in the example .git directory
path = Path('/home/foo/repo/home/foo/repo/.git/bar.py')
```

Reply:
Currently, my `is_relative_to` directly copies the python 3.9 implementation of `is_relative_to`, which I believe is the safest practice.

Reply:
How about:
```python
and not str(path).startswith(str(git_dir_path))
```
Seems like covers all edge cases?

Reply:
Solved. However, I just want to say that my personal philosophy is that if a functionality has a standard library implementation, we should try to follow that.

Reply:
@ashleve Please have a look 😄

Reply:
Thanks, I like your approach!
It's just that I believe this specific repo will be the most useful when it's as simple and clean as possible. I'm afraid that after taking many such design choices the code could just get too complex